File: frontend/lib/api/compare.ts
/**
 * lib/api/compare.ts — Edit 결과 ↔ 원본 비교 분석 호출.
 * 백엔드 POST /api/studio/compare-analyze 래퍼.
 *
 * Phase 6 (2026-04-27): 동기 JSON → task-based SSE 로 전환.
 * Phase 6 V4 (2026-05-05): compare context 응답 V4 (VisionCompareAnalysisV4 · 2-stage observe + diff_synthesize).
 *   - POST → {task_id, stream_url} 받음 → SSE drain → done event payload 추출
 *   - opts.onStage 콜백으로 stage 이벤트 실시간 전달 (PipelineTimeline 연동)
 *
 * SSE 5 stage 시퀀스 (compare context):
 *   compare-encoding → observe1 → observe2 → diff-synth → translation
 * Edit context 는 옛 시퀀스 유지 (intent-refine + vision-pair + translation).
 *
 * USE_MOCK 모드에선 stage emit + sleep 후 가짜 결과 반환 (실 백엔드 패턴 모사).
 */

import { STUDIO_BASE, USE_MOCK, parseSSE } from "./client";
import type { AnalyzeStageEvent } from "./vision";
import type { TaskCreated } from "./generated-helpers";
import { mockCompareAnalyze } from "./mocks/compare";
import type { ComparisonAnalysis, VisionCompareAnalysisV4 } from "./types";

export interface CompareAnalyzeRequest {
  /** 원본 이미지 / IMAGE_A — File / data URL / 절대 URL */
  source: File | string;
  /** 수정 결과 이미지 / IMAGE_B — File / data URL / 절대 URL */
  result: File | string;
  /** Edit context 일 때 사용자 수정 지시 (compare context 에선 무시). */
  editPrompt: string;
  /** 있으면 백엔드가 DB 에 영구 저장. tsk-{12hex} 형식. */
  historyItemId?: string;
  /** 비전 모델 override (기본 settings.visionModel). */
  visionModel?: string;
  /** 번역 모델 override (기본 settings.ollamaModel). */
  ollamaModel?: string;
  /**
   * 코드 경로 분기 (백엔드 default "edit" · 미전송 시 Edit 경로 = 기존 동작 100%).
   * "compare" → V4 pipeline (compare_pipeline_v4 · 2-stage observe + diff_synthesize)
   */
  context?: "edit" | "compare";
  /** Vision Compare 메뉴 전용 힌트 (context="compare" 일 때만 사용). */
  compareHint?: string;
  /** Phase 6 — stage 이벤트 도착 시 호출 (PipelineTimeline 의 stageHistory 갱신). */
  onStage?: (e: AnalyzeStageEvent) => void;
  /**
   * @deprecated spec §6.2 (2026-05-05) — V4 pipeline 은 promptMode 분기 없음.
   * 필드는 caller 호환을 위해 유지하지만 compareAnalyze 가 더 이상 백엔드로 보내지 않음.
   * 백엔드 v4 receiver 는 키 받아도 무시 (안전망).
   */
  promptMode?: "fast" | "precise";
}

export interface CompareAnalyzeResponse {
  /**
   * 분석 결과. context="edit" → ComparisonAnalysis (Edit v3 도메인 슬롯),
   * context="compare" → VisionCompareAnalysisV4 (V4 · 2-stage observe + diff_synthesize).
   * 호출자가 context 를 알고 있으므로 적절히 narrow 가능.
   */
  analysis: ComparisonAnalysis | VisionCompareAnalysisV4;
  /** historyItemId 가 DB 에 존재하고 갱신 성공 시 true. */
  saved: boolean;
}

function getStudioImageRef(input: File | string): string | null {
  if (input instanceof File) return null;
  if (input.startsWith("/images/")) return input;

  try {
    const url = new URL(input);
    const studioBase = new URL(STUDIO_BASE);
    if (url.origin === studioBase.origin && url.pathname.startsWith("/images/")) {
      return url.pathname;
    }
  } catch {
    return null;
  }

  return null;
}

/** File / data URL / 절대 URL → Blob 변환 (Edit 의 패턴 동일). */
async function toBlob(input: File | string): Promise<Blob> {
  // File 객체는 그대로 반환
  if (input instanceof File) return input;
  // 문자열(data URL / 절대 URL) 은 fetch 후 Blob 변환
  const res = await fetch(input);
  if (!res.ok) {
    throw new Error(`image fetch ${res.status}: ${input.slice(0, 80)}`);
  }
  return res.blob();
}

/**
 * 원본 · 수정 결과 이미지 쌍을 백엔드에 보내 비교 분석 결과를 가져온다.
 *
 * - Mock 모드: 800~1400ms sleep 후 가짜 ComparisonAnalysis 반환.
 * - Real 모드: multipart/form-data POST → shape 검증 → 반환.
 * - 응답 shape 검증: analysis 필드 존재 + 객체 타입 확인. 실패 시 Error throw.
 */
export async function compareAnalyze(
  req: CompareAnalyzeRequest,
): Promise<CompareAnalyzeResponse> {
  const isCompare = req.context === "compare";

  // Mock 분기 — UI 개발 시 실 백엔드 없이 동작 확인용 (stage emit 모사)
  if (USE_MOCK) {
    return mockCompareAnalyze(req);
  }

  // Real 분기: multipart/form-data 빌드
  const form = new FormData();
  const sourceRef = getStudioImageRef(req.source);
  const resultRef = getStudioImageRef(req.result);
  if (sourceRef) {
    form.append("source_ref", sourceRef);
  } else {
    const sourceBlob = await toBlob(req.source);
    form.append("source", sourceBlob, "source.png");
  }
  if (resultRef) {
    form.append("result_ref", resultRef);
  } else {
    const resultBlob = await toBlob(req.result);
    form.append("result", resultBlob, "result.png");
  }
  // meta 빌드 — context 없으면 백엔드 기본 "edit" 으로 동작 (기존 호출자 100% 무영향)
  const metaPayload: Record<string, unknown> = {
    editPrompt: req.editPrompt,
    historyItemId: req.historyItemId,
    visionModel: req.visionModel,
    ollamaModel: req.ollamaModel,
    sourceRef,
    resultRef,
  };
  if (isCompare) {
    metaPayload.context = "compare";
    metaPayload.compareHint = req.compareHint ?? "";
  }
  // spec §6.2 (2026-05-05): promptMode 더 이상 백엔드로 보내지 않음.
  // V4 pipeline 이 무관하고, Edit context 도 통일. caller 가 promptMode 넣어도 무시.
  form.append("meta", JSON.stringify(metaPayload));

  // Phase 6: POST 가 task_id + stream_url 반환 (옛처럼 직접 결과 X)
  const createRes = await fetch(`${STUDIO_BASE}/api/studio/compare-analyze`, {
    method: "POST",
    body: form,
  });
  if (!createRes.ok) {
    throw new Error(`compare-analyze failed: ${createRes.status}`);
  }
  // Tier 3 (2026-04-27): generated OpenAPI 타입 사용
  const { stream_url } = (await createRes.json()) as TaskCreated;

  // SSE drain — done event payload 추출. stage 이벤트는 onStage 콜백.
  const streamRes = await fetch(`${STUDIO_BASE}${stream_url}`, {
    headers: { accept: "text/event-stream" },
  });
  if (!streamRes.ok) {
    throw new Error(`compare-analyze stream ${streamRes.status}`);
  }

  for await (const evt of parseSSE(streamRes)) {
    if (evt.event === "error") {
      const payload = evt.data as { message?: string };
      throw new Error(payload.message || "compare-analyze error");
    }
    if (evt.event === "stage") {
      if (req.onStage) {
        const payload = evt.data as {
          type: string;
          progress: number;
          stageLabel: string;
        } & Record<string, unknown>;
        const { type, progress, stageLabel, ...extra } = payload;
        req.onStage({
          type,
          progress,
          stageLabel,
          extra: Object.keys(extra).length > 0 ? extra : undefined,
        });
      }
      continue;
    }
    if (evt.event === "done") {
      const json = evt.data as Partial<CompareAnalyzeResponse>;
      if (!json.analysis || typeof json.analysis !== "object") {
        throw new Error("compare-analyze: malformed done payload");
      }
      return {
        analysis: json.analysis,
        saved: !!json.saved,
      };
    }
  }
  throw new Error("compare-analyze: stream closed without done event");
}

/* ──────────────────────────────────────────────────────────────────────
 * V4 on-demand t2i prompt 합성 (Task 30 · 2026-05-05).
 * 메인 분석 결과의 observation1/2 중 하나로부터 prompt_synthesize 5 슬롯 합성.
 * 단일 JSON 응답 (non-SSE) — 약 10~20초 + GPU lock (busy 시 503).
 * ──────────────────────────────────────────────────────────────────── */

export interface PerImagePromptResponse {
  summary: string;
  positive_prompt: string;
  negative_prompt: string;
  key_visual_anchors: string[];
  uncertain: string[];
}

export async function compareAnalyzePerImagePrompt(
  observation: Record<string, unknown>,
  ollamaModel?: string,
): Promise<PerImagePromptResponse> {
  const res = await fetch(
    `${STUDIO_BASE}/api/studio/compare-analyze/per-image-prompt`,
    {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify({ observation, ollamaModel }),
    },
  );
  if (res.status === 503) {
    const body = (await res.json().catch(() => ({}))) as {
      detail?: { code?: string; message?: string };
    };
    throw new Error(
      body.detail?.message || "GPU busy — 잠시 후 다시 시도해주세요",
    );
  }
  if (!res.ok) {
    throw new Error(`per-image-prompt failed: ${res.status}`);
  }
  return res.json();
}

File: backend/studio/routes/compare.py
"""
studio.routes.compare — compare-analyze (Edit 결과 vs 원본 분석).

context 분기:
  edit (default): analyze_pair v3 — 도메인 분기 + 5 슬롯 매트릭스 + 의도 점수
  compare:       analyze_pair_v4 — Vision Compare 재설계 (관찰자→편집자 듀얼 + 5 카테고리 diff)

Phase 6 (2026-04-27): 동기 JSON 응답 → task-based SSE 로 전환.
  POST /compare-analyze              → { task_id, stream_url }
  GET  /compare-analyze/stream/{id}  → SSE (event: stage / done / error)

Task 12 (2026-05-05): per-image-prompt endpoint (단일 응답 · non-SSE).
  POST /compare-analyze/per-image-prompt → PerImagePromptResponse

GPU lock + ComfyUI 충돌 방지 정책은 백그라운드 파이프라인이 그대로 유지.
"""

from __future__ import annotations

import io

from fastapi import APIRouter, File, Form, HTTPException, Request, UploadFile
from fastapi.responses import StreamingResponse
from PIL import Image as PILImage

# Phase 6: 백그라운드 파이프라인 import (compare 도메인 로직 자체는 변경 없음)
from .._gpu_lock import GpuBusyError, gpu_slot
from ..comparison_pipeline import analyze_pair  # noqa: F401 — 옛 테스트 호환 (mock.patch 위치 · v3 edit context)
from ..pipelines import _run_compare_analyze_pipeline
from ..prompt_pipeline import clarify_edit_intent  # noqa: F401 — 옛 테스트 호환 (mock.patch 위치)
from ..schemas import PerImagePromptRequest, PerImagePromptResponse, TaskCreated
from ..storage import STUDIO_MAX_IMAGE_BYTES
from ..tasks import TASKS, _new_task
# Task 12 (V4): per-image t2i prompt endpoint — vision 정공법의 prompt_synthesize 재사용.
from ..vision_pipeline.prompt_synthesize import synthesize_prompt
from ._common import _spawn, _stream_task, parse_meta_object

router = APIRouter()


@router.post("/compare-analyze", response_model=TaskCreated)
async def create_compare_analyze_task(
    source: UploadFile = File(...),
    result: UploadFile = File(...),
    meta: str = Form(...),
):
    """비교 분석 요청 (multipart). Phase 6 — task 생성 + SSE.

    multipart:
      source: 원본 이미지 파일 (Vision Compare 컨텍스트에선 IMAGE_A)
      result: 결과 이미지 파일 (Vision Compare 컨텍스트에선 IMAGE_B)
      meta: JSON {context?, editPrompt?, compareHint?, historyItemId?, visionModel?, ollamaModel?}

    응답: { task_id, stream_url } — 클라이언트가 stream_url 로 SSE 구독.
    """
    meta_obj = parse_meta_object(meta)

    context = (meta_obj.get("context") or "edit").strip().lower()
    edit_prompt = (meta_obj.get("editPrompt") or "").strip()
    compare_hint = (meta_obj.get("compareHint") or "").strip()
    history_item_id_raw = meta_obj.get("historyItemId")
    vision_override = meta_obj.get("visionModel") or meta_obj.get("vision_model")
    text_override = meta_obj.get("ollamaModel") or meta_obj.get("ollama_model")
    # Phase 2 (2026-05-01) — Edit 자동 트리거 케이스에서 Edit 모드 패스스루.
    # 수동 Compare (Vision Compare 메뉴) 는 promptMode 미전달 → fast.
    compare_prompt_mode_raw = meta_obj.get("promptMode") or meta_obj.get("prompt_mode")
    compare_prompt_mode: str = (
        "precise" if isinstance(compare_prompt_mode_raw, str) and compare_prompt_mode_raw == "precise" else "fast"
    )

    source_bytes = await source.read()
    result_bytes = await result.read()
    if not source_bytes or not result_bytes:
        raise HTTPException(400, "empty image (source or result)")
    if (
        len(source_bytes) > STUDIO_MAX_IMAGE_BYTES
        or len(result_bytes) > STUDIO_MAX_IMAGE_BYTES
    ):
        raise HTTPException(413, "image too large")

    # ── Task 10 (V4): PIL verify + size 추출 ──
    # 옛 코드는 bytes 길이만 체크 → text 위장 / 손상 PNG 우회 가능했음.
    # V4 의 observe_image 가 width/height 인자를 요구 (per-image 슬롯 계산용).
    try:
        img_a = PILImage.open(io.BytesIO(source_bytes))
        img_a.verify()
        # verify() 후엔 stream 이 닫혀 size 접근 불가 → 재 open 필수
        img_a = PILImage.open(io.BytesIO(source_bytes))
        source_w, source_h = img_a.size
        if source_w <= 0 or source_h <= 0:
            raise ValueError("zero size")
    except Exception:
        raise HTTPException(400, "invalid image (source)")

    try:
        img_b = PILImage.open(io.BytesIO(result_bytes))
        img_b.verify()
        img_b = PILImage.open(io.BytesIO(result_bytes))
        result_w, result_h = img_b.size
        if result_w <= 0 or result_h <= 0:
            raise ValueError("zero size")
    except Exception:
        raise HTTPException(400, "invalid image (result)")

    task = await _new_task()
    task.worker = _spawn(
        _run_compare_analyze_pipeline(
            task,
            source_bytes=source_bytes,
            result_bytes=result_bytes,
            source_w=source_w,
            source_h=source_h,
            result_w=result_w,
            result_h=result_h,
            context=context,
            edit_prompt=edit_prompt,
            compare_hint=compare_hint,
            history_item_id_raw=history_item_id_raw,
            vision_override=vision_override,
            text_override=text_override,
            prompt_mode=compare_prompt_mode,
        )
    )
    return TaskCreated(
        task_id=task.task_id,
        stream_url=f"/api/studio/compare-analyze/stream/{task.task_id}",
    )


@router.get("/compare-analyze/stream/{task_id}")
async def compare_analyze_stream(task_id: str, request: Request):
    task = TASKS.get(task_id)
    if task is None:
        raise HTTPException(status_code=404, detail="task not found")
    return StreamingResponse(
        _stream_task(task, request),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "X-Accel-Buffering": "no",
        },
    )


# ── Task 12 (V4): per-image t2i prompt endpoint ──
# 메인 분석 결과의 observation1/2 중 하나로부터 t2i prompt 5 슬롯을 합성.
# on-demand 호출 (사용자가 결과 화면에서 클릭) — 단일 JSON 응답 (non-SSE).
# Pydantic 모델은 ..schemas 에 박제 (Task 13).
@router.post(
    "/compare-analyze/per-image-prompt",
    response_model=PerImagePromptResponse,
)
async def compare_per_image_prompt(req: PerImagePromptRequest) -> PerImagePromptResponse:
    """observation JSON → t2i prompt 합성 (단일 응답 · non-SSE).

    on-demand 호출 — 메인 분석 후 사용자가 결과 화면에서 클릭. 약 10~20초.
    gpu_slot('compare-per-image-prompt') 안에서 호출 → busy 시 503 + code=gpu_busy.
    """
    if not req.observation:
        raise HTTPException(400, "empty observation")

    text_model = req.ollamaModel or "gemma4-un:latest"
    try:
        async with gpu_slot("compare-per-image-prompt"):
            synth = await synthesize_prompt(
                req.observation,
                text_model=text_model,
                timeout=60.0,
                ollama_url="http://localhost:11434",
            )
    except GpuBusyError as e:
        raise HTTPException(
            status_code=503,
            detail={"code": "gpu_busy", "message": str(e)},
        )

    return PerImagePromptResponse(
        summary=synth.get("summary", "") or "",
        positive_prompt=synth.get("positive_prompt", "") or "",
        negative_prompt=synth.get("negative_prompt", "") or "",
        key_visual_anchors=synth.get("key_visual_anchors", []) or [],
        uncertain=synth.get("uncertain", []) or [],
    )
